src/System.Private.CoreLib/shared/System/Index.cs

```diff
@@ -0,0 +1,27 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System
+{
+    public readonly struct Index
+    {
+        private readonly int _value;
+
+        public int Value => _value < 0 ? ~_value : _value;
+        public bool FromEnd => _value < 0;
+
+        public Index(int value, bool fromEnd)
+        {
+            if (value < 0)
+            {
+                throw new ArgumentException(SR.ArgumentOutOfRange_NeedNonNegNum, nameof(value));
+            }
+
+            _value = fromEnd ? ~value : value;
+        }
+
+        public static implicit operator Index(int value)
+            => new Index(value, fromEnd: false);
```

Comment:
Should these types have overridden Equals/GetHashCode/ToString like other core structs ?

Reply:
LDM hasn't talked about this that I can find, but I think it's a good idea.

Reply:
I'll add these. any suggestion  for ToString return values? 

I assume for Index we should return the index and ^ for from end, something like "1" and "^1"

For Range, we can return x..y and add ^ before the number if any of the indexes is from end. 

what you think?

Reply:
I think that sounds good. cc @MadsTorgersen I'll bring this to LDM for confirmation, but you're good to put it in for preview 1.

Reply:
Should they implement `IEquatable<T>`?

Reply:
I'll go ahead and implement it.